src/ess/powder/transform.py

```diff
@@ -0,0 +1,69 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+"""Signal transformation algorithms for powder diffraction."""
+
+import scipp as sc
+
+from .uncertainty import UncertaintyBroadcastMode, broadcast_uncertainties
+
+
+def compute_pdf_from_structure_factor(
+    s: sc.DataArray,
+    *,
+    rho: sc.Variable,
+    r: sc.Variable,
+    uncertainty_broadcast_mode=UncertaintyBroadcastMode.drop,
+) -> sc.DataArray:
+    '''
+    Compute a pair distribution function from a structure factor.
+
+    Computes the pair distribution function :math:`D(r)` as defined in
+    `Review: Pair distribution functions from neutron total scattering for the study of local structure in disordered materials <https://www.sciencedirect.com/science/article/pii/S2773183922000374>`_
+    from the overall scattering function :math:`S(Q)`.
+
+    The inputs to the algorithm are:
+
+    * A histogram representing :math:`S(Q)` with :math:`N` bins on a bin-edge grid with :math:`N+1` edges :math:`Q_j` for :math:`j=0\\ldots N`.
+    * The density :math:`\\rho`, the total number of atoms per volume.
+    * The bin-edge grid over :math:`r` the output histogram representing :math:`D(r)` will be computed on.
+
+    In each output bin, the output is computed as:
+
+    .. math::
+        D_{i+\\frac{1}{2}} &= \\frac{2}{\\pi(r_{i+1}-r_i)} \\int_{r_i}^{r_{i+1}} \\int_{0}^\\infty (S(Q) - 1) Q \\sin(Q r) dQ \\ dr  \\\\
+        &\\approx \\frac{2}{\\pi(r_{i+1}-r_i)} \\sum_{j=0}^{N-1} (S(Q)_{j+\\frac{1}{2}} - 1) (\\cos(\\bar{Q}_{j+\\frac{1}{2}} r_{i})-\\cos(\\bar{Q}_{j+\\frac{1}{2}} r_{i+1})) \\Delta Q_{j+\\frac{1}{2}}
+
+    Note that in the above expression the subscript :math:`_{j+\\frac{1}{2}}` is used to denote
+    quantities belonging to the :math:`j^\\text{th}` bin of a histogram, :math:`\\bar{Q}_{j+\\frac{1}{2}} = \\frac{Q_j + Q_{j+1}}{2}` and :math:`\\Delta Q_{j+\\frac{1}{2}} = Q_{j+1} - Q_{j}`.
+
+    Parameters
+    ----------
+    s:
+        1D DataArray representing :math:`S(Q)` with bin-edge coordinate :math:`Q`
+    rho:
+        total number of atoms per unit volume
+    r:
+        1D array, bin-edges of output grid
+    uncertainty_broadcast_mode: UncertaintyBroadcastMode,
+        Choose how uncertainties in S(Q) are broadcast to D(r).
+        Defaults to ``UncertaintyBroadcastMode.drop``.
```

Comment:
I think `fail` is the better, more defensive default. I'd rather do the safer alternative by default and require intervention to use a riskier option.

Reply:
Won't that make the function fail 100% of the time unless the user manually specifies the `uncertainty_broadcast_mode` to be something different? Seems excessively safe and not very useful.

The application of a linear operator on a vector will cause the entries of the output vector to be correlated, the only way we can avoid correlation in this case is by not invoking the function.

If we want the user to be aware that the output data is not uncorrelated maybe it's better to display a warning.

Reply:
Warnings are almost always ignored. Only hard errors get people's attention.

Everywhere else, the default is to raise on variance broadcasts. So why is this function different?

> Won't that make the function fail 100% of the time unless the user manually specifies the uncertainty_broadcast_mode to be something different?

Or when users remove variances early. But in practice, people may never want to drop them but use the upper bound method instead.

Reply:
I think this kind of protecting the users from themselves is a mistake. This makes the user having to specify the `uncertainty_propagation_mode` argument every time this function is called (if the input has variances, which I think is the common case). That is an annoying UX.

We must assume the users know to some extent what they are doing. This function does not support (the classical type of) uncertainty propagation. Of course that does not mean we should return erroneous or misleading results, that's why I'm suggesting making it the default to `drop` variances.

Dropping the variances will not cause erroneous results. Worst case scenario it causes an error later when the result doesn't have variances but the user expected it to have them. I think that is better than the user having to specify the `uncertainty_propagation_mode` argument every time this function is called.

Reply:
I am inclined to say that if the result has no error bars at all, then making that the default is ok. If the output *does* have error bars, but some *contribution* to them is dropped, doing that by default is problematic.